lib/spack/llnl/util/filesystem.py

```diff
@@ -497,8 +582,115 @@ def basenames(self):
 
         >>> l = LibraryList(['/dir1/liba.a', '/dir2/libb.a', '/dir3/liba.a'])
         >>> assert l.basenames == ['liba.a', 'libb.a']
+        >>> h = HeaderList(['/dir1/a.h', '/dir2/b.h', '/dir3/a.h'])
+        >>> assert h.basenames == ['a.h', 'b.h']
+        """
+        return list(dedupe(os.path.basename(x) for x in self.files))
+
+    @property
+    def names(self):
+        """Stable de-duplication of file names in the list
+
+        >>> h = HeaderList(['/dir1/a.h', '/dir2/b.h', '/dir3/a.h'])
+        >>> assert h.names == ['a', 'b']
+        """
+        return list(dedupe(x.split('.')[0] for x in self.basenames))
+
+    def __getitem__(self, item):
+        cls = type(self)
+        if isinstance(item, numbers.Integral):
+            return self.files[item]
+        return cls(self.files[item])
+
+    def __add__(self, other):
+        return self.__class__(dedupe(self.files + list(other)))
+
+    def __radd__(self, other):
+        return self.__add__(other)
+
+    def __eq__(self, other):
+        return self.files == other.files
+
+    def __len__(self):
+        return len(self.files)
+
+    def joined(self, separator=' '):
+        return separator.join(self.files)
+
+    def __repr__(self):
+        return self.__class__.__name__ + '(' + repr(self.files) + ')'
+
+    def __str__(self):
+        return self.joined()
+
+
+class HeaderList(FileList):
+    """Sequence of absolute paths to headers
+
+    Provides a few convenience methods to manipulate header paths and get
+    commonly used compiler flags or names
+    """
+
+    @property
+    def headers(self):
+        return self.files
+
+    @property
+    def cpp_flags(self):
+        """Include flags
+
+        >>> h = HeaderList(['/dir1/a.h', '/dir1/b.h', '/dir2/c.h'])
+        >>> assert h.cpp_flags == '-I/dir1 -I/dir2'
         """
-        return list(dedupe(os.path.basename(x) for x in self.libraries))
+        return ' '.join(['-I' + x for x in self.directories])
+
+
+def find_headers(headers, root, recurse=False):
+    """Returns an iterable object containing a list of full paths to
+    headers if found.
+
+    Accepts any glob characters accepted by fnmatch:
```

Comment:
New features, yay! This allows you to search for headers like `*.h` or libraries like `*.so`.